ext/QuadGKEnzymeExt.jl

```diff
@@ -0,0 +1,100 @@
+
+module QuadGKEnzymeExt
+
+using QuadGK, Enzyme, LinearAlgebra
+
+function Enzyme.EnzymeRules.augmented_primal(config, ofunc::Const{typeof(quadgk)}, ::Type{RT}, f, segs::Annotation{T}...; kws...) where {RT, T<:Real}
+    prims = map(x->x.val, segs)
+
+    retres, segbuf = if f isa Const
+        if EnzymeRules.needs_primal(config)
+            quadgk(f.val, prims...; kws...), nothing
+        else
+            nothing
+        end
+    else
+        I, E, segbuf = quadgk_segbuf(f.val, prims...; kws...)
+        if EnzymeRules.needs_primal(config)
+            (I, E), segbuf
+        else
+            nothing, segbuf
+        end
+    end
+
+    dres = if !Enzyme.EnzymeRules.needs_shadow(config)
+        nothing
+    elseif EnzymeRules.width(config) == 1
+        zero.(res...)
+    else
+        ntuple(Val(EnzymeRules.width(config))) do i
+            Base.@_inline_meta
+            zero.(res...)
+        end
+    end
+
+    cache = if RT <: Duplicated || RT <: DuplicatedNoNeed || RT <: BatchDuplicated || RT <: BatchDuplicatedNoNeed
+        dres
+    else
+        nothing
+    end
+    cache2 = segbuf, cache
+
+    return Enzyme.EnzymeRules.AugmentedReturn{
+        Enzyme.EnzymeRules.needs_primal(config) ? eltype(RT) : Nothing,
+        Enzyme.EnzymeRules.needs_shadow(config) ? (Enzyme.EnzymeRules.width(config) == 1 ? eltype(RT) : NTuple{Enzyme.EnzymeRules.width(config), eltype(RT)}) : Nothing,
+        typeof(cache2)
+    }(retres, dres, cache2)
+end
+
+function call(f, x)
+    f(x)
+end
+
+struct ClosureVector{F}
+    f::F
+end
+
+@inline function guaranteed_nonactive(::Type{T}) where T
+    rt = Enzyme.Compiler.active_reg_inner(T, (), nothing)
+    return rt == Enzyme.Compiler.AnyState || rt == Enzyme.Compiler.DupState
+end
+
+function Base.:+(a::ClosureVector, b::ClosureVector)
+    Enzyme.Compiler.recursive_add(a, b, identity, guaranteed_nonactive)
+end
```

Comment:
Shouldn't it have to call this on `a.f` and `b.f`, and wrap the result in a `ClosureVector`?  Or does `Enzyme.Compiler.recursive_add` do this for you?

Also, is this type-stable?  i.e. will the integrand at every point x be a `ClosureVector{F}` of the same type `F`, and will `Enzyme.Compiler.recursive_add` preserve this type?

Reply:
Since recursive_add deals with arbitrary types already using Closurevector directly is equally good as a.f. It does have the property that the inputs and result types are equivalent/type preserving. It may or may not be type stable (I think it should be though), but we can explicitly cast it to ::typeof(a) on return if desired

Reply:
Would be good to check type stability here — if we sum up the vJp at 1000s of quadrature points, it would be bad if it creates 2000 new `ClosureVector` types (one for every `+`, `-`, and `*` operation)!